src/go/k8s/internal/testutils/k3s.go

```diff
@@ -0,0 +1,605 @@
+	stack.initializeK3s(t)
+	stack.initializeHelm(t)
+	stack.installCertManagerChart(t)
+
+	stack.portForwarder = newPortForwarder(stack.rest, stack.client)
+	return stack
+}
+
+func (k *K3sStack) InstallCluster(t *testing.T, options RedpandaOptions) *v1alpha2.Redpanda {
+	id := k.installations.Add(1)
+	options = options.normalize(id)
+	k.installRedpandaChart(t, options)
+
+	jsonData, err := json.Marshal(options.Values)
+	require.NoError(t, err)
+
+	cluster := &v1alpha2.Redpanda{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      options.Name,
+			Namespace: options.Namespace,
+		},
+		Spec: v1alpha2.RedpandaSpec{
+			ClusterSpec: &v1alpha2.RedpandaClusterSpec{},
+		},
+	}
+	err = json.Unmarshal(jsonData, &cluster.Spec.ClusterSpec)
+	require.NoError(t, err)
+
+	k.mutex.Lock()
+	defer k.mutex.Unlock()
+
+	k.clusters[id] = cluster
+	return cluster
+}
+
+func (k *K3sStack) Client() client.Client {
+	return k.client
+}
+
+type portForwarder struct {
+	config *rest.Config
+	client client.Client
+}
+
+func newPortForwarder(config *rest.Config, client client.Client) *portForwarder {
+	return &portForwarder{
+		config: config,
+		client: client,
+	}
+}
+
+func (p *portForwarder) freePort() (port int, err error) {
+	addr, err := net.ResolveTCPAddr("tcp", "localhost:0")
+	if err != nil {
+		return 0, err
+	}
+
+	listener, err := net.ListenTCP("tcp", addr)
+	if err != nil {
+		return 0, err
+	}
+	defer listener.Close()
+
+	return listener.Addr().(*net.TCPAddr).Port, nil
+}
+
+func getPortFromSpec(spec *v1alpha2.RedpandaClusterSpec, name string) int {
+	defaultPort := defaultPorts[name]
+	if spec == nil || spec.Listeners == nil {
+		return defaultPort
+	}
+	switch name {
+	case "admin":
+		if spec.Listeners.Admin != nil && spec.Listeners.Admin.Port != nil {
+			return *spec.Listeners.Admin.Port
+		}
+	case "http":
+		if spec.Listeners.HTTP != nil && spec.Listeners.HTTP.Port != nil {
+			return *spec.Listeners.HTTP.Port
+		}
+	case "kafka":
+		if spec.Listeners.Kafka != nil && spec.Listeners.Kafka.Port != nil {
+			return *spec.Listeners.Kafka.Port
+		}
+	case "rpc":
+		if spec.Listeners.RPC != nil && spec.Listeners.RPC.Port != nil {
+			return *spec.Listeners.RPC.Port
+		}
+	case "schemaRegistry":
+		if spec.Listeners.SchemaRegistry != nil && spec.Listeners.SchemaRegistry.Port != nil {
+			return *spec.Listeners.SchemaRegistry.Port
+		}
+	}
+	return defaultPort
+}
+
+type forwardedPorts struct {
+	mappings map[string]string
+	cancel   context.CancelFunc
+	wg       *sync.WaitGroup
+	stopped  atomic.Bool
+}
+
+func (p *forwardedPorts) Stop() {
+	p.stopped.Store(true)
+	p.cancel()
+	p.wg.Wait()
+}
+
+func (p *forwardedPorts) IsStopped() bool {
+	return p.stopped.Load()
+}
+
+func (p *portForwarder) forwardClusterPorts(ctx context.Context, cluster *v1alpha2.Redpanda, name string) (*forwardedPorts, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	targetURL, err := url.Parse(p.config.Host)
+	if err != nil {
+		cancel()
```

Comment:
stylistic nit: I think you'd be better off using named return values here so there can be a single defer call to conditionally cancel the context instead of having to remember to call `cancel()` at all the early returns.

```go
func name() (_*forwardedPorts, err error) {
  ctx, cancel := context.WithCancel(ctx)
  
  defer func() {
    if err != nil {
      cancel()
    }
  }()
}
```